### ISSUE TITLE
the trait bound uuid::Uuid: diesel::Expression is not satisfied

### ISSUE THREAD

**Issue by frederikhors** [diesel-rs/diesel#2348]
### Versions

- **Rust:** 1.42.0
- **Diesel:** 1.4.4
- **Database:** Postgres 12
- **Operating System:** Win 10

### Cargo.toml
```
diesel = { version = "=1.4.4", features = ["postgres", "r2d2", "uuid", "chrono"] }
diesel_migrations = "=1.4.0"
...
uuid = { version = "=0.8.1", features = ["serde", "v4"] }
```

### What are you trying to accomplish?

Use `uuid` with `diesel` with no problems.

### What is the expected output?

It works.

### What is the actual output?

`the trait bound uuid::Uuid: diesel::Expression is not satisfied`

### Are you seeing any additional errors?

Nope.

I saw:

- https://github.com/uuid-rs/uuid/issues/453 and
- https://github.com/diesel-rs/diesel/pull/2235

But still problems, **why?**

## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues) for similar issues.
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)

**Comment by weiznich** [comment]
Basically a duplicate of https://github.com/diesel-rs/diesel/issues/1900.

(You need to use the `uuidv07` feature to make this work.)

**Comment by frederikhors** [comment]
@weiznich thank for your answer.

Now I'm using:

```toml
[dependencies]
actix-web = "=2.0.0"
actix-rt = "=1.0.0"
chrono = { version = "=0.4.11", features = ["serde"] }
diesel = { version = "=1.4.4", features = ["postgres", "r2d2", "uuidv07", "chrono"] }
diesel_migrations = "=1.4.0"
dotenv = "=0.15.0"
env_logger = "=0.7.1"
lazy_static = "=1.4.0"
log = "=0.4.8"
r2d2 = "=0.8.8"
serde = "=1.0.105"
serde_json = "=1.0.50"
uuid = { version = "=0.8.1", features = ["serde", "v4"] }
```

But Cargo gives me a warning:

`warning: Feature uuidv07 of package diesel has been obsolete in version 1.4.4`

**Comment by weiznich** [comment]
We do not say that anywhere in our code, so I'm not sure where cargo is getting this information from. You should probably rise a issue there.

**Comment by frederikhors** [comment]
`there` where?

**Comment by weiznich** [comment]
In the issue tracker of whichever tool is giving you this warning. So if it's cargo itself then here: https://github.com/rust-lang/cargo/, if it's some custom subcommand in then in their issue tracker.

**Comment by problame** [comment]
FWIW, I am using diesel 1.4.5 with uuid 0.8.1  and adding the `uuidv07` feature just fixed the problem for me. Is there a cfg! in diesel 1.4.5 that still uses `uuidv07` ?

**Comment by gaku-sei** [comment]
Second @problame's issue, the following dependencies and features doesn't work:

```toml
diesel = {version = "1.4.4", features = ["chrono", "postgres", "r2d2", "uuid"]}
uuid = "0.8.2"
```

But this does:

```toml
diesel = {version = "1.4.4", features = ["chrono", "postgres", "r2d2", "uuidv07"]}
uuid = "0.8.2"
```

**Comment by HopefulFire** [comment]
Wow this should be reopened/ a new issue created. Same problem for me, was fixed by @gaku-sei 's solution

**Comment by enokson** [comment]
I just had the same issue. The issue was also resolved by @gaku-sei 's solution.

**Comment by An7One** [comment]
Unfortunately, none works for mine.

```
[dependencies]
actix-cors = "0.5.4"
actix-rt = "1.1.1"
actix-service = "1.0.6"
actix-web = "3.3.2"
bcrypt = "0.10.1"
chrono = { version = "0.4.19", features = ["serde"] }
derive_more = "0.99.11"
diesel = {version = "1.4.4", features = ["chrono", "postgres", "r2d2", "uuidv07"]}
diesel_migrations = "1.4.0"
dotenv = "0.15.0"
env_logger = "0.9.0"
# for heroku rust buildpack
libsqlite3-sys = { version = "0.22.2", features = ["bundled"] }
failure = "0.1.8"
futures = "0.3.12"
jsonwebtoken = "7.2.0"
log = "0.4.14"
serde = "1.0.123"
serde_derive = "1.0.123"
serde_json = "1.0.62"
# uuid = { version = "0.8.1", features = ["serde", "v4"]}
uuid = "0.8.2"
```

**Comment by An7One** [comment]
Other things that were helpful to me:
1. to match order [src](https://github.com/diesel-rs/diesel/issues/2011#issuecomment-472375646)
2. to deserialize nullable type [src](https://github.com/diesel-rs/diesel/issues/330#issuecomment-219279105):

**Comment by KodrAus** [comment]
Is there a trait impl we could add over in `uuid` that would clear all this up? We’re currently working on stabilizing `uuid` so hopefully can smooth things out for `diesel` one way or another.

**Comment by weiznich** [comment]
@KodrAus Thanks for reaching out here ❤️ 
There are a few common uuid related issues in diesel:

* Mixing up `diesel::sql_types::Uuid` and `uuid::Uuid`. This is basically an issue with users [ignoring our documentation](https://docs.rs/diesel/1.4.8/diesel/#serializing-and-deserializing), which explicitly states that types from `diesel::sql_types` cannot hold values and shouldn't be used anywhere in your "model". This is something that we can only improve on diesels side.
* For the 1.4.x release series there seems to be some confusion about which feature flag enables support for uuid. There is the "uuid" and the "uuid_v07" feature, where the former enables the support for `uuid` < 0.7 and the later for the 0.7 and 0.8 releases. As of today I personally believe it was a mistake to name that features that way, but unfortunately there is no way to reverse that. Fortunately, the next diesel release will be 2.0, so we can fix that there. It's already better on the master branch, where we've just dropped support for uuid < 0.7 and only have single uuid feature left. Maybe I will also rename the features for third party crates to indicate the version numbers in a better way, so that this does not happen anymore. So in the end, again that's something to fix on diesels side.

That written: The uuid integration in diesel is really simple, as it's [only a few lines of code](https://github.com/diesel-rs/diesel/blob/b2c58897c39c519a946314bd5b63765d3af56204/diesel/src/pg/types/uuid.rs). As long as the two function we are using are not  changed in an incompatible way there is really not much maintained burden here. Having a stable uuid 1.0 release should really help there.

**Comment by Corallus-Caninus** [comment]
I'm getting the trait 'diesel::Expression' is not implemented for 'uuid::Uuid' on 1.4.8 and uuid 8.2 with feature uuid in diesel.

**Comment by Corallus-Caninus** [comment]
I fixed with using the diesel::*::Uuid type in schema and uuid::Uuid crate in model and my application.

**Comment by KodrAus** [comment]
Thanks for the details @weiznich!

Yeh, there's no way those `Uuid::from_fields` and `Uuid::as_bytes` will ever change. So your next release is going to be `2.0`? Depending on how you want to go, if you wanted to consolidate your `uuid` features to just its `1.0` you should be safe depending on `uuid`'s `1.0.0-alpha.1` release. We're treating it as already stable, and will follow up with a `1.0.0` proper once it's proven itself. Cargo should happily consider `1.0.0-alpha.1` and `1.0.0` as semver compatible.

**Comment by weiznich** [comment]
@KodrAus Thanks for that information. Yes that's the plan to update the dependency to `1.0` before doing the diesel 2.0 release.

**Comment by piscespieces** [comment]
The solution referenced above works

```
diesel = {version = "1.4.8", features = ["uuidv07", "r2d2", "postgres"]}
```

note the `"uuidv07"` instead of just `"uuid"`

**Comment by KodrAus** [comment]
Just a heads up @weiznich I'm running through that `1.0.0` release of `uuid` now: https://github.com/uuid-rs/uuid/pull/596

**Comment by weiznich** [comment]
@KodrAus Thanks for the heads up. I've pushed https://github.com/diesel-rs/diesel/pull/3137 to add support for uuid 1.0.0 to diesel master :tada:

**Comment by CuriousCI** [comment]
Shouldn't the ```1.4.x``` release  support ```uuid 1.0.0``` too? I'm currently facing some issues with Rocket.rs as it relies on diesel 1.4, but it's uuid feature is implemented for ```uuid 1.0.0```. I'll open an issue to add ```2.0.0``` support for diesel in Rocket, but until then I have to monkey patch the problem: I downgraded to ```uuid 0.8```, and manually implemented Rocket's traits.

**Comment by weiznich** [comment]
@CuriousCI We don't plan to release any additional 1.4.x release at this point in time as we just don't have the resources to maintain more than one diesel version in parallel.

**Comment by tadman** [comment]
[`uuid`](https://crates.io/crates/uuid) is at 1.1.2 now so any dependencies which depend on `uuid` might get ahead of Diesel and foul things up.

`cargo tree` helps track down errant dependencies, but this is a short-term solution. Hopefully Diesel can bump to at least `1.1.x` soon.

**Comment by weiznich** [comment]
@tadman Please do not repeat questions that are already asked above. To repeat that: We won't add support for uuid 1.0 to any old release. Support for this uuid version will be only there on the upcoming 2.0 release. Please use the rc release if you need to have it now.

**Comment by jayvdb** [comment]
I ran into this error, and struggled for a bit to understand the problem.  IMO it would be very helpful to add a example which uses `Uuid`, `chrono::*`, etc.  I am happy to do this, if others agree.
(Sorry @weiznich for posting this on the wrong issue.)

**Comment by weiznich** [comment]
@jayvdb That's the last warning: Stop posting to old issues. If you have questions: Use the discussion forum. If you have suggestions: Write a proposal in the discussion forum, but do not post on old issues and do not ping people.

**Comment by ashupednekar** [comment]
the `uuidv07` feature no longer exists

```bash
error: unrecognized feature for crate diesel: uuidv07
disabled features:
    128-column-tables, 64-column-tables, __with_asan_tests, extras, huge-tables
    i-implement-a-third-party-backend-and-opt-into-breaking-changes, ipnet-address
    large-tables, mysql, mysql_backend, mysqlclient-src, network-address, numeric
    pq-src, quickcheck, returning_clauses_for_sqlite_3_35, sqlite, time, unstable
    without-deprecated
enabled features:
    32-column-tables, chrono, postgres, postgres_backend, r2d2, serde_json, uuid
    with-deprecated
```

**Comment by weiznich** [comment]
@ashupednekar Commenting to an old issue is the wrong way to ask for support. Please open a new thread in our [support forum](https://github.com/diesel-rs/diesel/discussions) outlining your problem. This should include a minimal reproducible example and a description of your problem.

**Comment by ashupednekar** [comment]
Got it, will do
